add remote debug option for otel-collector to builder

adamkubon opened the pull request:
**Description:** add option to enable remote debug

**Link to tracking Issue:** Closes https://github.com/open-telemetry/opentelemetry-collector/issues/6149

**Testing:** Tested with delve

**Documentation:** Documentation updated

jpkrohling commented on cmd/builder/README.md:
```diff
@@ -65,6 +65,17 @@ $ ocb --config=builder-config.yaml
 
+## Debug
+
+To enable debug for otel-collector-binary set configuration flag `debug-compilation` to true.
```
```suggestion
To keep the debug symbols in the resulting OpenTelemetry Collector binary, set the configuration property `debug_compilation` to true.
```

jpkrohling commented on cmd/builder/README.md:
```diff
@@ -65,6 +65,17 @@ $ ocb --config=builder-config.yaml
+
+To enable debug for otel-collector-binary set configuration flag `debug-compilation` to true.
+
+Then install `go-delve` and run otel-collector with `dlv` command as following example:
```
```suggestion
Then install `go-delve` and run OpenTelemetry Collector with `dlv` command as the following example:
```

jpkrohling commented on cmd/builder/README.md:
````diff
@@ -65,6 +65,17 @@ $ ocb --config=builder-config.yaml
+# go install github.com/go-delve/delve/cmd/dlv@latest
+# ~/go/bin/dlv --listen=:2345 --headless=true --api-version=2 --accept-multiclient --log exec .otel-collector-binary -- --config otel-collector-config.yaml
+```
+Finally load otel-collector as a project into IDE, configure debug for Go
````
```suggestion
Finally, load the OpenTelemetry Collector as a project in the IDE, configure debug for Go
```

Side note: how to do that? Do you have a link to an external document on how to "load as a project in the IDE"?

jpkrohling commented on cmd/builder/README.md:
```diff
@@ -90,6 +101,7 @@ dist:
     output_path: /tmp/otelcol-distributionNNN # the path to write the output (sources and binary). Optional.
     version: "1.0.0" # the version for your custom OpenTelemetry Collector. Optional.
     go: "/usr/bin/go" # which Go binary to use to compile the generated sources. Optional.
+    debug-compilation: true # this option enable debugging possibility with go-delve tool. Optional.
```
```suggestion
    debug_compilation: false # enabling this causes the builder to keep the debug symbols in the resulting binary. Optional.
```

Where possible, we show the default value here.

jpkrohling commented on cmd/builder/internal/builder/config.go:
```diff
@@ -56,6 +56,7 @@ type Distribution struct {
 	OutputPath     string `mapstructure:"output_path"`
 	Version        string `mapstructure:"version"`
 	BuildTags      string `mapstructure:"build_tags"`
+	Debug          bool   `mapstructure:"debug_compilation"`
```
```suggestion
	DebugSymbols          bool   `mapstructure:"debug_compilation"`
```

jpkrohling commented on cmd/builder/internal/command.go:
```diff
@@ -110,7 +111,7 @@ configuration is provided, ocb will generate a default Collector.
 		return nil, err
 	}
-
+	cmd.Flags().BoolVar(&cfg.Distribution.Debug, distributionDebugFlag, false, "Compile with debug flags (default false)")
```
We are moving away from flags in favor of config files only. It might be better not to add a new flag at this point.

adamkubon replied:
removed

jpkrohling commented on cmd/builder/internal/builder/main.go:
```diff
@@ -87,6 +87,9 @@ func Compile(cfg Config) error {
 	cfg.Logger.Info("Compiling")
 
 	args := []string{"build", "-ldflags=-s -w", "-trimpath", "-o", cfg.Distribution.Name}
+	if cfg.Distribution.DebugSymbols {
```
I would prefer to have the contents of ldflags to be a var that is used in the args list. This way, the list can be the same independent of the flag's value, instead of repeating the same list, just to change the ldflags arg.

adamkubon replied:
done

jpkrohling commented on cmd/builder/internal/command.go:
```diff
@@ -40,6 +40,7 @@ const (
 	distributionOutputPathFlag     = "output-path"
 	distributionGoFlag             = "go"
 	distributionModuleFlag         = "module"
+	distributionDebugFlag          = "debug-compilation"
```
Shouldn't this be removed?

adamkubon replied:
removed

adamkubon commented on cmd/builder/internal/command.go:
```diff
@@ -192,4 +192,7 @@ func applyCfgFromFile(flags *flag.FlagSet, cfgFromFile builder.Config) {
 	if !flags.Changed(distributionModuleFlag) && cfgFromFile.Distribution.Module != "" {
 		cfg.Distribution.Module = cfgFromFile.Distribution.Module
 	}
+	if !flags.Changed(distributionDebugFlag) && cfgFromFile.Distribution.DebugSymbols != false {
```
removed

bogdandrutu commented:
@jpkrohling PTAL

jpkrohling reviewed (commented):
There's something odd, can you confirm this is indeed working? `readelf` shows the exact same output for two binaries, each with a different `debug_compilation` value:

```
$ readelf -S /tmp/otelcol-distribution954067720/otelcol-custom
There are 27 section headers, starting at offset 0x270:

Section Headers:
  [Nr] Name              Type             Address           Offset
       Size              EntSize          Flags  Link  Info  Align
  [ 0]                   NULL             0000000000000000  00000000
       0000000000000000  0000000000000000           0     0     0
  [ 1] .text             PROGBITS         0000000000401000  00001000
       000000000087ef05  0000000000000000  AX       0     0     32
  [ 2] .plt              PROGBITS         0000000000c7ff20  0087ff20
       0000000000000280  0000000000000010  AX       0     0     16
  [ 3] .rodata           PROGBITS         0000000000c81000  00881000
       0000000000385f2c  0000000000000000   A       0     0     32
  [ 4] .rela             RELA             0000000001006f30  00c06f30
       0000000000000018  0000000000000018   A      11     0     8
  [ 5] .rela.plt         RELA             0000000001006f48  00c06f48
       00000000000003a8  0000000000000018   A      11     2     8
  [ 6] .gnu.version_r    VERNEED          0000000001007300  00c07300
       0000000000000050  0000000000000000   A      10     1     8
  [ 7] .gnu.version      VERSYM           0000000001007360  00c07360
       0000000000000058  0000000000000002   A      11     0     2
  [ 8] .hash             HASH             00000000010073c0  00c073c0
       00000000000000d4  0000000000000004   A      11     0     8
  [ 9] .shstrtab         STRTAB           0000000000000000  00c074a0
       0000000000000105  0000000000000000           0     0     1
  [10] .dynstr           STRTAB           00000000010075c0  00c075c0
       0000000000000245  0000000000000000   A       0     0     1
  [11] .dynsym           DYNSYM           0000000001007820  00c07820
       0000000000000420  0000000000000018   A      10     1     8
  [12] .typelink         PROGBITS         0000000001007c40  00c07c40
       0000000000006f6c  0000000000000000   A       0     0     32
  [13] .itablink         PROGBITS         000000000100ebc0  00c0ebc0
       0000000000002d88  0000000000000000   A       0     0     32
  [14] .gosymtab         PROGBITS         0000000001011948  00c11948
       0000000000000000  0000000000000000   A       0     0     1
  [15] .gopclntab        PROGBITS         0000000001011960  00c11960
       000000000059cab0  0000000000000000   A       0     0     32
  [16] .go.buildinfo     PROGBITS         00000000015af000  011af000
       0000000000001990  0000000000000000  WA       0     0     16
  [17] .dynamic          DYNAMIC          00000000015b09a0  011b09a0
       0000000000000120  0000000000000010  WA      10     0     8
  [18] .got.plt          PROGBITS         00000000015b0ac0  011b0ac0
       0000000000000150  0000000000000008  WA       0     0     8
  [19] .got              PROGBITS         00000000015b0c10  011b0c10
       0000000000000008  0000000000000008  WA       0     0     8
  [20] .noptrdata        PROGBITS         00000000015b0c20  011b0c20
       000000000004bb98  0000000000000000  WA       0     0     32
  [21] .data             PROGBITS         00000000015fc7c0  011fc7c0
       0000000000017c70  0000000000000000  WA       0     0     32
  [22] .bss              NOBITS           0000000001614440  01214440
       000000000003a7e0  0000000000000000  WA       0     0     32
  [23] .noptrbss         NOBITS           000000000164ec20  0124ec20
       0000000000013950  0000000000000000  WA       0     0     32
  [24] .tbss             NOBITS           0000000000000000  00000000
       0000000000000008  0000000000000000 WAT       0     0     8
  [25] .interp           PROGBITS         0000000000400fe4  00000fe4
       000000000000001c  0000000000000000   A       0     0     1
  [26] .note.go.buildid  NOTE             0000000000400f80  00000f80
       0000000000000064  0000000000000000   A       0     0     4
Key to Flags:
  W (write), A (alloc), X (execute), M (merge), S (strings), I (info),
  L (link order), O (extra OS processing required), G (group), T (TLS),
  C (compressed), x (unknown), o (OS specific), E (exclude),
  D (mbind), l (large), p (processor specific)
```

It would be good to have a positive confirmation in the build logs stating that the debug symbols will be kept when that option is set.

adamkubon commented:
> 

Fixed - I removed too much in when I was applying the review fixes and the debug option was not passed to the config. Now the readelf -S will show the different output for the both cases and during compilation additional info will inform about Enabling remote debug.

It was caused because of lack any tests for the command.go file - I'll try to add some tests to it, to cover the case.

adamkubon commented on cmd/builder/internal/builder/config.go:
```diff
@@ -56,6 +56,7 @@ type Distribution struct {
 	OutputPath     string `mapstructure:"output_path"`
 	Version        string `mapstructure:"version"`
 	BuildTags      string `mapstructure:"build_tags"`
+	DebugSymbols   bool   `mapstructure:"debug_compilation"`
```
renamed

adamkubon commented:
Rebased
